transports/quic/src/endpoint.rs

```diff
@@ -100,75 +90,71 @@ pub struct Endpoint {
 
-    /// Socketaddr of the local UDP socket passed in the configuration at initialization after it
-    /// has potentially been modified to handle port number `0`.
-    pub(crate) local_addr: SocketAddr,
+    socket_addr: SocketAddr,
 }
 
 impl Endpoint {
-    /// Builds a new `Endpoint`.
-    pub fn new(config: Config) -> Result<Arc<Endpoint>, io::Error> {
-        let local_socket_addr = match crate::transport::multiaddr_to_socketaddr(&config.multiaddr) {
-            Some(a) => a,
-            None => panic!(), // TODO: Err(TransportError::MultiaddrNotSupported(multiaddr)),
-        };
+    /// Builds a new [`Endpoint`] that is listening on the [`SocketAddr`].
+    pub fn new_bidirectional(
+        config: Config,
+        socket_addr: SocketAddr,
+    ) -> Result<(Arc<Endpoint>, mpsc::Receiver<Connection>), transport::Error> {
+        let (new_connections_tx, new_connections_rx) = mpsc::channel(1);
+        let endpoint = Self::new(config, socket_addr, Some(new_connections_tx))?;
+        Ok((endpoint, new_connections_rx))
+    }
 
-        // NOT blocking, as per man:bind(2), as we pass an IP address.
-        let socket = std::net::UdpSocket::bind(&local_socket_addr)?;
-        // TODO:
-        /*let port_is_zero = local_socket_addr.port() == 0;
-        let local_socket_addr = socket.local_addr()?;
-        if port_is_zero {
-            assert_ne!(local_socket_addr.port(), 0);
-            assert_eq!(multiaddr.pop(), Some(Protocol::Quic));
-            assert_eq!(multiaddr.pop(), Some(Protocol::Udp(0)));
-            multiaddr.push(Protocol::Udp(local_socket_addr.port()));
-            multiaddr.push(Protocol::Quic);
-        }*/
+    /// Builds a new [`Endpoint`] that only supports outbound connections.
+    pub fn new_dialer(config: Config) -> Result<Arc<Endpoint>, transport::Error> {
+        let socket_addr = SocketAddrV4::new(Ipv4Addr::LOCALHOST, 0);
+        Self::new(config, socket_addr.into(), None)
+    }
 
+    fn new(
+        config: Config,
+        socket_addr: SocketAddr,
+        new_connections: Option<mpsc::Sender<Connection>>,
+    ) -> Result<Arc<Endpoint>, transport::Error> {
+        // NOT blocking, as per man:bind(2), as we pass an IP address.
+        let socket = std::net::UdpSocket::bind(&socket_addr)?;
         let (to_endpoint_tx, to_endpoint_rx) = mpsc::channel(32);
         let to_endpoint2 = to_endpoint_tx.clone();
-        let (new_connections_tx, new_connections_rx) = mpsc::channel(1);
 
         let endpoint = Arc::new(Endpoint {
             to_endpoint: Mutex::new(to_endpoint_tx),
             to_endpoint2,
-            new_connections: Mutex::new(new_connections_rx),
-            local_addr: socket.local_addr()?,
+            socket_addr: socket.local_addr()?,
         });
 
+        let server_config = new_connections.map(|c| (c, config.server_config.clone()));
+
         // TODO: just for testing, do proper task spawning
         async_global_executor::spawn(background_task(
```

Comment:
@kpp:
> I believe this one can be polled directly, right?

Reply:
If I understand you correctly, you are proposing to have the logic of the `background_task` directly in `Listener::poll_next` instead of a separate task, right?

I considered doing this, but after an out of band discussion with @mxinden decided against it. 
The reason for this is that every time that any event / reading / writing / udp packet on any connection happens, the endpoint is involved and thus the waker of its task will be called. `Listener::poll` is driven by `Transport::poll`, which in return is polled directly in the `Swarm`'s main loop that also drives the behaviours and the `Pool`. Thus they all share the same waker. Right now if one of these structures triggers the waker, all of them are polled again. If we drive the endpoint is this main loop, this would create unnecessary overhead. 
Does that explanation make sense?  
I agree that this is not ideal; alternative ideas would be appreciated.